src/snowflake/cli/plugins/nativeapp/project_model.py

```diff
@@ -0,0 +1,182 @@
+# Copyright (c) 2024 Snowflake Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from functools import cached_property
+from pathlib import Path
+from typing import List, Optional
+
+from snowflake.cli.api.cli_global_context import cli_context
+from snowflake.cli.api.project.definition import (
+    default_app_package,
+    default_application,
+    default_role,
+)
+from snowflake.cli.api.project.schemas.native_app.application import (
+    ApplicationPostDeployHook,
+)
+from snowflake.cli.api.project.schemas.native_app.native_app import NativeApp
+from snowflake.cli.api.project.schemas.native_app.path_mapping import PathMapping
+from snowflake.cli.api.project.util import extract_schema, to_identifier
+from snowflake.cli.plugins.nativeapp.artifacts import resolve_without_follow
+from snowflake.connector import DictCursor
+
+
+def current_role() -> str:
```

Comment:
I believe this can be used incorporated in https://github.com/snowflakedb/snowflake-cli/blob/851bfdbda0cb4a9521c46fc4c17273adb36fbd36/src/snowflake/cli/api/sql_execution.py#L91 to reduce the duplicate logic in that context manager.

Reply:
Spotted that too, but we don't have ownership where it should go, so I'd prefer to split it into its own submission to avoid noise for the CLI team.

Reply:
TBH I think it's worth getting their sign-off here, it's so generally useful. But we can always make a one-off PR afterwards that moves it.

Reply:
Yeah, I don't think the current_role() calls belongs there, I was always planning to move it due to it being generally useful. There's a location that has all these default wh/role/etc helpers, I'm planning to move it there too.